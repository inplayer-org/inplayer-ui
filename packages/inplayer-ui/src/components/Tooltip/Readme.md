Tooltip a is a component for displaying additional information on hover some element. The position of the tooltip attribute can be: up, down, right, left, up-right, up-left, down-left and down-right. You can use it directly like this:

```jsx static
import { Tooltip } from 'inplayer-ui';

function Page(props) {
  return (
    <Tooltip
      content="ðŸ˜Ž"
    >
      <Button>
        Hover me !
      </Button>
    </Tooltip>
  );
}
```

### Demo

```jsx
<Tooltip
  content="ðŸ˜Ž"
>
  <Button>
    Hover me!
  </Button>
</Tooltip>
<br />
<br />
<Tooltip
  content="Tooltip at the left"
  placement="left"
>
  <Button>
    Hover me too!
  </Button>
</Tooltip>
<br />
<br />
<Tooltip
  content="Tooltip at the right"
  placement="right"
>
  <Button>
    Hover me as well!
  </Button>
</Tooltip>
<br />
<br />
<Tooltip
  content={`Tooltip at the bottom\nTooltip at the bottom`}
  placement="bottom"
>
  <Button>
    Hover me also!
  </Button>
</Tooltip>
<br />
<br />
<Tooltip
  content="Hello!"
  behavior="click"
  placement="right"
>
  <Button>
    Click me!
  </Button>
</Tooltip>
```

### Tooltip with Custom Color

```jsx
<Tooltip
  arrow={15}
  background="#000"
  border="#000"
  color="#fff"
  content="ðŸ˜Ž"
  fadeDuration={0}
  fadeEasing="linear"
  fixed={false}
  fontFamily="inherit"
  fontSize="inherit"
  offset={0}
  padding={2}
  placement="top"
  radius={0}
  zIndex={1}
>
  <Button>
    Hover me !
  </Button>
</Tooltip>
```
