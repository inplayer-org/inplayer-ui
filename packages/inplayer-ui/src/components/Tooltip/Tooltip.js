// @flow
import React from 'react';
import styled, { css, keyframes } from 'styled-components';

export type FadeEasing = 'linear' | 'ease' | 'ease-in' | 'ease-out' | 'ease-in-out';

type Props = {
  children: Node,
  offset: number,
  isOpen: boolean,
  placement: string,
  zIndex: number,
  fadeEasing: FadeEasing,
  fadeDuration: number,
};

const fadeAnimation = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

const animation = props => css`
  animation: ${props.fadeDuration}ms ${props.fadeEasing} 0s 1 ${fadeAnimation};
`;

const Base = styled('div')`
  position: absolute;
  width: max-content;
  white-space: pre-line;
  ${props => props.fadeDuration && props.fadeDuration > 0 && animation(props)};
  ${props => props.zIndex && `z-index: ${props.zIndex};`};
`;

const Top = styled(Base)`
  bottom: 100%;
  left: 50%;
  transform: translateX(-50%);
  margin-bottom: ${props => props.offset}px;
`;

const Bottom = styled(Base)`
  top: 100%;
  left: 50%;
  transform: translateX(-50%);
  margin-top: ${props => props.offset}px;
`;

const Left = styled(Base)`
  right: 100%;
  top: 50%;
  transform: translateY(-50%);
  margin-right: ${props => props.offset}px;
`;

const Right = styled(Base)`
  left: 100%;
  top: 50%;
  transform: translateY(-50%);
  margin-left: ${props => props.offset}px;
`;

const tooltips = {
  left: Left,
  top: Top,
  right: Right,
  bottom: Bottom,
};

const Tooltip = ({
  children,
  offset,
  isOpen,
  placement,
  zIndex,
  fadeDuration,
  fadeEasing,
  ...props
}: Props) => {
  const Component = tooltips[placement] || tooltips.top;

  if (!isOpen) {
    return null;
  }

  return (
    <Component
      offset={offset}
      zIndex={zIndex}
      fadeDuration={fadeDuration}
      fadeEasing={fadeEasing}
      {...props}
    >
      {children}
    </Component>
  );
};

export default Tooltip;
