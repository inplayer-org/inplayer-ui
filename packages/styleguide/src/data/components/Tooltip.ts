import { ComponentData } from "../types";

const usage = `
  import { Tooltip } from 'inplayer-ui';

  function Page(props) {
    return (
      <Tooltip
        content="ðŸ˜Ž"
      >
        <Button>
          Hover me !
        </Button>
      </Tooltip>
    );
  }
`;

const demoCode = `<div>
  <Tooltip
    content="ðŸ˜Ž"
  >
    <Button>
      Hover me!
    </Button>
  </Tooltip>
  <br />
  <br />
  <Tooltip
    content="Tooltip at the left"
    placement="left"
  >
    <Button>
      Hover me too!
    </Button>
  </Tooltip>
  <br />
  <br />
  <Tooltip
    content="Tooltip at the right"
    placement="right"
  >
    <Button>
      Hover me as well!
    </Button>
  </Tooltip>
  <br />
  <br />
  <Tooltip
    content={\`Tooltip at the bottom\nTooltip at the bottom\`}
    placement="bottom"
  >
    <Button>
      Hover me also!
    </Button>
  </Tooltip>
  <br />
  <br />
  <Tooltip
    content="Hello!"
    behavior="click"
    placement="right"
  >
    <Button>
      Click me!
    </Button>
  </Tooltip>
</div>`;

const tooltipCustomColor = `<Tooltip
  arrow={15}
  background="#000"
  border="#000"
  color="#fff"
  content="ðŸ˜Ž"
  fadeDuration={0}
  fadeEasing="linear"
  fixed={false}
  fontFamily="inherit"
  fontSize="inherit"
  offset={0}
  padding={2}
  placement="top"
  radius={0}
  zIndex={1}
  >
  <Button>
    Hover me !
  </Button>
</Tooltip>
`;

const Tooltip: ComponentData = {
  description: `
    Tooltip a is a component for displaying additional information on hover some element. The position of the tooltip attribute can be: up, down, right, left, up-right, up-left, down-left and down-right. You can use it directly like this:
  `,
  path: 'src/components/Tooltip/index.js',
  usage,
  examples: [
    {
      title: 'Demo',
      code: demoCode,
    },
    {
      title: 'Tooltip with Custom Color',
      code: tooltipCustomColor,
    }
  ],
  propsAndMethods: [
    { name: 'content', type: 'Node', default: 'Required' },
    { name: 'arrowWidth', type: 'number	', default: '8' },
    { name: 'background', type: 'string' },
    { name: 'behavior', type: 'enum', default: 'hover', description: 'One of: hover, click, ref' },
    { name: 'border', type: 'string' },
    { name: 'children', type: 'Node', default: 'null' },
    { name: 'className', type: 'string', default: 'null' },
    { name: 'color', type: 'string' },
    { name: 'durationOnClick', type: 'number', default: '1000'},
    { name: 'fadeDuration', type: 'number', default: '150'},
    { name: 'fadeEasing', type: 'FadeEasing', default: 'linear'},
    { name: 'fixed', type: 'boolean', default: 'false'},
    { name: 'fontFamily', type: 'string', default: 'inherit'},
    { name: 'fontSize', type: 'string', default: 'null'},
    { name: 'offset', type: 'number', default: '0'},
    { name: 'padding', type: 'number', default: '0.7'},
    { name: 'placement', type: 'enum', default: 'top', description: 'One of: left, right, top, bottom' },
    { name: 'radius', type: 'number', default: '0'},
    { name: 'style', type: 'object', default: '{ }'},
    { name: 'zIndex', type: 'number', default: '1'},
  ],
}

export default Tooltip;