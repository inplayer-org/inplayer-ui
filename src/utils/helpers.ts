import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
