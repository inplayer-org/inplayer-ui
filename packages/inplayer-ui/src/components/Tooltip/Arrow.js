// @flow
import React from 'react';
import styled from 'styled-components';
import { ifProp, prop } from 'styled-tools';

import { uiColors } from 'utils';

type Props = {
  background: string,
  border: string,
  placement: string,
  width: number,
};

const Base = styled('div')`
  position: absolute;
  width: ${props => props.width}px;
  height: ${props => props.width}px;
  background: ${ifProp('background', prop('background'), uiColors('secondary.main'))};
`;

const Up = styled(Base)`
  transform: translateX(-50%) translateY(50%) rotateZ(45deg);
  bottom: 100%;
  left: 50%;
  border-left: 1px solid ${props => props.border};
  border-top: 1px solid ${props => props.border};
`;

const Down = styled(Base)`
  transform: translateX(-50%) translateY(-50%) rotateZ(45deg);
  top: 100%;
  left: 50%;
  border-right: 1px solid ${props => props.border};
  border-bottom: 1px solid ${props => props.border};
`;

const Left = styled(Base)`
  transform: translateX(50%) translateY(-50%) rotateZ(45deg);
  right: 100%;
  top: 50%;
  border-left: 1px solid ${props => props.border};
  border-bottom: 1px solid ${props => props.border};
`;

const Right = styled(Base)`
  transform: translateX(-50%) translateY(-50%) rotateZ(45deg);
  left: 100%;
  top: 50%;
  border-right: 1px solid ${props => props.border};
  border-top: 1px solid ${props => props.border};
`;

const arrows = {
  left: Right,
  top: Down,
  right: Left,
  bottom: Up,
};

const Arrow = ({ background, border, placement, width }: Props) => {
  const Component = arrows[placement] || arrows.top;
  return width > 0 && <Component background={background} border={border} width={width} />;
};

export default Arrow;
