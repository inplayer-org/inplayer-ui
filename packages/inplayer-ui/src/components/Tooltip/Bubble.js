import styled from 'styled-components';
import { ifProp, prop } from 'styled-tools';
import colors from 'config/colors';
import { uiColors, fontSizes } from 'utils';

const Bubble = styled('div')`
  color: ${colors.white};
  background: ${ifProp('background', prop('background'), uiColors('secondary.main'))};
  border-radius: ${props => (props.radius ? `${props.radius}px` : '0.3em')};
  border: 1px solid ${props => props.border};
  padding: ${props => (props.padding ? `${props.padding}rem` : 0)};
  font-size: ${props => props.fontSize || fontSizes('small')};
  font-family: ${props => props.fontFamily};
  line-height: 1.4;
`;

export default Bubble;
