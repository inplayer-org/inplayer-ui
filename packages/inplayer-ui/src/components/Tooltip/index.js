// @flow
import React, { PureComponent } from 'react';
import styled from 'styled-components';

import Arrow from './Arrow';
import TooltipWrapper from './Tooltip';
import Bubble from './Bubble';
import type { FadeEasing } from './Tooltip';

export type Placement = 'left' | 'right' | 'top' | 'bottom';
export type TooltipBehavior = 'hover' | 'click' | 'ref';

export type Props = {
  behavior?: TooltipBehavior,
  durationOnClick?: number,
  arrowWidth?: number,
  background?: string,
  border?: string,
  children?: Node,
  color?: string,
  content: Node,
  className?: string,
  style?: object,
  fadeDuration?: number,
  fadeEasing?: FadeEasing,
  fixed?: boolean,
  fontFamily?: string,
  fontSize?: string,
  offset?: number,
  padding?: number,
  placement?: Placement,
  radius?: number,
  zIndex?: number,
};

type State = {
  isOpen: boolean,
};

const Container = styled.div`
  position: relative;
  display: inline-block;
`;

class Tooltip extends PureComponent<Props, State> {
  state = {
    isOpen: false,
  };

  timeDelay = (msDuration: number) =>
    new Promise(resolve => setTimeout(() => resolve(true), msDuration));

  showTooltip = () => {
    this.setState({ isOpen: true });
  };

  hideTooltip = () => {
    this.setState({ isOpen: false });
  };

  flashTooltip = async () => {
    const { durationOnClick } = this.props;
    const { isOpen } = this.state;

    if (isOpen) {
      return;
    }

    this.showTooltip();
    await this.timeDelay(durationOnClick);
    this.hideTooltip();
  };

  render() {
    const { isOpen } = this.state;
    const {
      behavior,
      arrowWidth,
      background,
      border,
      children,
      color,
      content,
      style,
      className,
      fadeDuration,
      fadeEasing,
      fixed,
      fontFamily,
      fontSize,
      offset,
      padding,
      placement,
      radius,
      zIndex,
    } = this.props;

    const tooltipElement = (
      <TooltipWrapper
        isOpen={!children || fixed ? true : isOpen}
        placement={placement}
        offset={offset + arrowWidth}
        zIndex={zIndex}
        fadeEasing={fadeEasing}
        fadeDuration={fadeDuration}
      >
        <Bubble
          background={background}
          border={border}
          color={color}
          radius={radius}
          fontFamily={fontFamily}
          fontSize={fontSize}
          padding={padding}
        >
          <Arrow width={arrowWidth} background={background} border={border} placement={placement} />
          {content}
        </Bubble>
      </TooltipWrapper>
    );

    return children ? (
      <Container
        style={style}
        className={className}
        onClick={behavior === 'click' ? this.flashTooltip : undefined}
        onMouseEnter={behavior === 'hover' && !fixed ? this.showTooltip : undefined}
        onMouseLeave={behavior === 'hover' && !fixed ? this.hideTooltip : undefined}
      >
        {children}
        {tooltipElement}
      </Container>
    ) : (
      <Container className={className} style={style}>
        {tooltipElement}
      </Container>
    );
  }
}

Tooltip.defaultProps = {
  behavior: 'hover',
  durationOnClick: 1000,
  arrowWidth: 8,
  background: '',
  border: '',
  children: null,
  color: '',
  fadeDuration: 150,
  fadeEasing: 'linear',
  fixed: false,
  fontFamily: 'inherit',
  fontSize: null,
  offset: 0,
  padding: 0.7,
  placement: 'top',
  radius: 0,
  zIndex: 1,
  style: {},
  className: null,
};

export default Tooltip;
