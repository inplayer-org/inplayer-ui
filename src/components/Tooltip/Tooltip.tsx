import React, {
  useState,
  ReactChild,
  ReactNode,
  forwardRef,
  useImperativeHandle,
  Ref,
} from 'react';
import styled from 'styled-components';
import Arrow from './Arrow';
import TooltipWrapper, { FadeEasing } from './TooltipWrapper';
import Bubble from './Bubble';
import { AnalyticsProps } from '../../analytics';

export interface FlashTooltipRef {
  flashTooltip: () => Promise<void>;
}

export type Placement = 'left' | 'right' | 'top' | 'bottom';
export type TooltipBehavior = 'hover' | 'click' | 'ref';

export type TooltipProps = {
  behavior?: TooltipBehavior;
  durationOnClick?: number;
  arrowWidth?: number;
  background?: string;
  color?: string;
  content?: ReactNode;
  fadeEasing?: FadeEasing;
  placement?: Placement;
  fadeDuration?: number;
  border?: string;
  fixed?: boolean;
  fontSize?: string;
  offset?: number;
  padding?: number;
  radius?: number;
  zIndex?: number;
  className?: string;
  children?: ReactChild | ReactNode | null;
  ref: Ref<FlashTooltipRef>;
} & AnalyticsProps;

const Container = styled.div`
  position: relative;
  display: inline-block;
`;

const Tooltip = forwardRef<FlashTooltipRef, TooltipProps>(
  (
    {
      behavior = 'hover',
      arrowWidth = 8,
      background = '',
      border = '',
      children = null,
      color = '',
      content,
      fadeDuration = 150,
      fadeEasing = 'linear',
      fixed = false,
      fontSize = '',
      offset = 0,
      padding = 0.7,
      placement = 'top',
      radius = 0,
      zIndex = 1,
      className = '',
      durationOnClick = 1000,
    },
    ref
  ) => {
    const [isOpen, setIsOpen] = useState(false);

    const timeDelay = (msDuration: number) =>
      new Promise((resolve) => setTimeout(() => resolve(true), msDuration));

    const showTooltip = () => setIsOpen(true);

    const hideTooltip = () => setIsOpen(false);

    const flashTooltip = async () => {
      if (isOpen) return;

      showTooltip();
      await timeDelay(durationOnClick);
      hideTooltip();
    };

    useImperativeHandle(ref, () => ({
      flashTooltip,
    }));

    const tooltipElement = (
      <TooltipWrapper
        isOpen={!children || fixed ? true : isOpen}
        placement={placement}
        offset={offset + arrowWidth}
        zIndex={zIndex}
        fadeEasing={fadeEasing}
        fadeDuration={fadeDuration}
      >
        <Bubble
          background={background}
          border={border}
          color={color}
          radius={radius}
          fontSize={fontSize}
          padding={padding}
        >
          <Arrow width={arrowWidth} background={background} border={border} placement={placement} />
          {content}
        </Bubble>
      </TooltipWrapper>
    );

    return (
      <Container
        className={className}
        onClick={behavior === 'click' ? flashTooltip : undefined}
        onMouseEnter={behavior === 'hover' && !fixed ? showTooltip : undefined}
        onMouseLeave={behavior === 'hover' && !fixed ? hideTooltip : undefined}
      >
        <>
          {children}
          {tooltipElement}
        </>
      </Container>
    );
  }
);

Tooltip.displayName = 'Tooltip';

export default Tooltip;
