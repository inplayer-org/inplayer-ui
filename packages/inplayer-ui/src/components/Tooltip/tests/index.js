import { generateSnapshots } from 'packages/inplayer-ui/src/components/Tooltip/tests/helpers';

import Tooltip from '../index';

generateSnapshots(Tooltip);
