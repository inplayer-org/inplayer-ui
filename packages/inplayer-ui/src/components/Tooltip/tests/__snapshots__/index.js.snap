// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders correctly 1`] = `
.c3 {
  position: absolute;
  width: 8px;
  height: 8px;
  background: #286090;
  -webkit-transform: translateX(-50%) translateY(-50%) rotateZ(45deg);
  -ms-transform: translateX(-50%) translateY(-50%) rotateZ(45deg);
  transform: translateX(-50%) translateY(-50%) rotateZ(45deg);
  top: 100%;
  left: 50%;
  border-right: 1px solid;
  border-bottom: 1px solid;
}

.c1 {
  position: absolute;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  white-space: pre-line;
  -webkit-animation: 150ms linear 0s 1 bcCCNc;
  animation: 150ms linear 0s 1 bcCCNc;
  z-index: 1;
  bottom: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  margin-bottom: 8px;
}

.c2 {
  color: #fff;
  background: #286090;
  border-radius: 0.3em;
  border: 1px solid;
  padding: 0.7rem;
  font-size: 0.875rem;
  font-family: inherit;
  line-height: 1.4;
}

.c0 {
  position: relative;
  display: inline-block;
}

<div
  className="c0"
  style={Object {}}
>
  <div
    className="c1"
    offset={8}
  >
    <div
      className="c2"
      color=""
      fontFamily="inherit"
      fontSize={null}
      radius={0}
    >
      <div
        className="c3"
        width={8}
      />
    </div>
  </div>
</div>
`;
